### PR TITLE
SN-14, SN-13, SN-110, SN-97, SN-111 Implementation, Bug Fix, Refactoring

### PR DESCRIPTION
**Pull request by deeppatel-11**
This Pull Request includes:
- Implementation: Notification.py allows users to be notified when mailing package, person, or user pet is Infront of the camera 
- Implementation: Using the coco dataset model, the images can be processed so that a .txt file is created automatically, which can later be used for Yolov5 training (it used to be done manually before this).
- Bug Fix: The camera fix by increasing delay.
- Refactoring by removing dead code, method extraction, class extraction, Renaming variables.

### CODE REVIEW

File: WebService/camera.py [hunk 1]
```diff
@@ -28,25 +33,26 @@ def get_frame(self):
...
+        # reading the mail-bird.names, mail-bird.names file. The coco model will help to detect dogs, cats,
+        # person or bird
+        mail_bird_classes = self.pet_detection.getClasses('names_files/mail-bird.names')
 
         # get the height and width
         height, width, _ = frame.shape
         blob = cv2.dnn.blobFromImage(frame, 1 / 255, (416, 416), (0, 0, 0), swapRB=True, crop=False)
 
-        #self.net.setInput(blob)
+        # self.net.setInput(blob)
```
Comment (hasan-ahmed) [inline WebService/camera.py:48]: this can be removed completely instead of being commented out

**Comment by hasan-ahmed** [comment]
**Variables Name.**
File name suggestion: notification.py can potentially be changed to notification_service to have consistency with other services (s3_service). 

**Functions**

Function get_frame has grown longer than ideal and is responsible for many different aspects. I will approve this regardless because I believe it is best to create an entire separate PR for the refactoring of the camera.py class in next sprint.

**PR Checklist**

- [ x]  Documentation present for every component and function.
- [ ]  Unit tests are implemented.
- [ x]  Implementation matches the corresponding requirements.
- [ x]  All commits linked to Jira ticket.

**Overall Remarks.**

Feature works on my local machine and code is clean. However, important to make sure to refactor the longer functions in future sprints and add unit tests.

**Comment by hasan-ahmed** [comment]
**Variables Name.**
File name suggestion: notification.py can potentially be changed to notification_service to have consistency with other services (s3_service). 

**Functions**

Function get_frame has grown longer than ideal and is responsible for many different aspects. I will approve this regardless because I believe it is best to create an entire separate PR for the refactoring of the camera.py class in next sprint.

**PR Checklist**

- [ x]  Documentation present for every component and function.
- [ ]  Unit tests are implemented.
- [ x]  Implementation matches the corresponding requirements.
- [ x]  All commits linked to Jira ticket.

**Overall Remarks.**

Feature works on my local machine and code is clean. However, important to make sure to refactor the longer functions in future sprints and add unit tests.

**Review by hasan-ahmed — APPROVED** [review]
**Variables Name.**
File name suggestion: notification.py can potentially be changed to notification_service to have consistency with other services (s3_service). 

**Functions**

Function get_frame has grown longer than ideal and is responsible for many different aspects. I will approve this regardless because I believe it is best to create an entire separate PR for the refactoring of the camera.py class in next sprint.

**PR Checklist**

- [ x]  Documentation present for every component and function.
- [ ]  Unit tests are implemented.
- [ x]  Implementation matches the corresponding requirements.
- [ x]  All commits linked to Jira ticket.

**Overall Remarks.**

Feature works on my local machine and code is clean. However, important to make sure to refactor the longer functions in future sprints and add unit tests.

**Comment by deeppatel-11** [comment]
> **Variables Name.** File name suggestion: notification.py can potentially be changed to notification_service to have consistency with other services (s3_service).
> 
> **Functions**
> 
> Function get_frame has grown longer than ideal and is responsible for many different aspects. I will approve this regardless because I believe it is best to create an entire separate PR for the refactoring of the camera.py class in next sprint.
> 
> **PR Checklist**
> 
> * [ x]  Documentation present for every component and function.
> * [ ]   Unit tests are implemented.
> * [ x]  Implementation matches the corresponding requirements.
> * [ x]  All commits linked to Jira ticket.
> 
> **Overall Remarks.**
> 
> Feature works on my local machine and code is clean. However, important to make sure to refactor the longer functions in future sprints and add unit tests.

File: WebService/bounding_boxes.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
+import cv2
+import numpy as np
+from pet_detection import petDetection
+import math
+import sys
+import matplotlib.pyplot as plt
+import os
+
+def convert(size, box):
```
Comment (MikeG3) [inline WebService/bounding_boxes.py:9]: function name suggestion: getBoxDimensions()

File: WebService/bounding_boxes.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        print("No object found")
+        sys.exit()
+
+    for i in indexes_coco.flatten():
+        label = []
+        bb = []
+        if str(coco_classes[class_ids_coco[i]]) == 'cat' or str(coco_classes[class_ids_coco[i]]) == 'dog':
+            pd.draw_bounding_boxes(boxes=boxes_coco, index=i, classes=coco_classes, class_ids=class_ids_coco,
+                                   confidences=confidences_coco, my_img=my_img, color=(0, 255, 0), labels=label)
+            # This stores the sze of each bounding box into a dictionary
```
Comment (MikeG3) [inline WebService/bounding_boxes.py:44]: Long Function suggestion:

Break this up into a sub-function: Ie. displayBoxInfo(x, y, w, h, type)

File: WebService/bounding_boxes.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        for f in os.listdir(newpath):
+            index = f.find('.')
+            textfile_name = f[:index]
+            print("FileName :" , textfile_name )
+            my_img = cv2.imread(newpath + '/' + f)
+            bb = analyze(my_img, net_coco, coco_classes, pd)
+            if not(bb == []):
+                file = open(newpath + '/' + textfile_name + '.txt', "a")
+                file.write(str(counter) + " ")
+                for b in range(len(bb)-1):
```
Comment (MikeG3) [inline WebService/bounding_boxes.py:85]: variable name suggestion: b --> boxIndex (*if Im understanding this code correctly)

File: WebService/camera.py [hunk 5]
```diff
@@ -6,20 +6,25 @@
...
 import math
 from pet_detection import petDetection
+from notification import notification
 
 
 class Video(object):
     def __init__(self, queueSize=128, cameraSource=0, timeout=0):
         self.video = CamVideoStream(0)
-        time.sleep(5)
+        time.sleep(10)
```
Comment (MikeG3) [inline WebService/camera.py:15]: Good! Video-delay on my end, can sometimes reach 10 seconds

**Review by MikeG3 — APPROVED** [review]
### Variables Name
bounding_boxes.py - line 85 5 --> suggestion: bounding_box_index

### Function Name


### Long Function
bounding_boxes.py - line 22 - analyze
Its a long function. A suggestion would be to split line 45 to 54 as a new function: display_box_dimension(boxes_coco[i])

### Code Complexity.
bounding_boxes.py - line 8 - conert
Suggestion: getBoxDimensions(size, box)

### PR Checklist

- [x] Documentation present for every component and function.
- [x] Unit tests are implemented.
- [x] Implementation matches the corresponding requirements.
- [x] All commits linked to Jira ticket.

### Overall Remarks.
Great work!!

**Comment by abirshah** [comment]
### _Variables Name._

In camera.py:
Line 42 variables name _mail_bird_classes should_ explain more about the variables. It can be improved to 
_indentifying_maling_package_and_bird_in_cat's_mouth_.
Line 39 variables name _coco_classes_ can be improved to _object_classes_from_coco_dataset._
Line 46 variables _blob_ doesn't really give implications about these variables.
LIne 54 variables _boxes_mail_bird, confidences_mail_bird, class_ids_mail_bird_ can be improved to 
_bounding_boxes_of_mailing_package_and_bird_in_cats_mouth_ and _confidence_socre_mailing_package_and_bird_in_cats_mouth_ also
_class_ids_mailing_package_bird_in_cats_mouth_
Line 118 variable colors can be ragactor to _displaying_different_colors_for_each_bounding_box_

### _Function Name_

In notification.py files
Line 9 _sendNotificaion_ can be improved to _sendEmailNotificatonForUserDetectedPet_
LIne 16 login can be improved to _loginToGamilService_

### _Long Function._

We should take refactor function in the camera.py files. _self. video.start()_ can be broken into 
multiple functions in a way that each function is responsible for only one task. and we can call 
functions in one function.


### _Code Complexity._

Code is very easily readable. SonarCoud is also giving good results.

image


### _PR Checklist_

  - [x] Documentation present for every component and function.
  - [] Unit tests are implemented.
  - [x] Implementation matches the corresponding requirements.
  - [x] All commits linked to Jira ticket.
  
### _Overall Remarks._

I will approve this PR. Refactoring in notifaction.py is excellent.

**Review by abirshah — APPROVED** [review]

**Pull request merged**